client/state/data-layer/wpcom/account-recovery/lookup/index.js

```diff
@@ -32,12 +35,15 @@ export const validate = ( { primary_email, primary_sms, secondary_email, seconda
 	}
 };
 
-export const handleRequestResetOptions = ( { dispatch }, { userData } ) => (
+export const handleRequestResetOptions = ( { dispatch }, { userData }, next, transit = page ) => (
```

Comment:
Driving by here — I've also had difficulty using the data layer for anything more complex than "Request some data and stuff it into the store". Similar to this case, for Happychat I needed to "request a token from the API at `/jwt/sign` and then connect Happychat with that token data." But since other pieces of the app might also need their own token, we can't trigger the HC connection from the data-layer handler itself.

I ended up abandoning this work, but my plan was essentially to ensure that some "context" data was sent with the `TOKEN_RECEIVE` action so that Happychat middleware could listen for all `TOKEN_RECEIVE`s but only respond to the one it cares about. For example:

```
// Happychat middleware needs a token before it can connect chat.
// It dispatches this action:
{
  type: 'TOKEN_REQUEST',
  signature: happychatTokenSignature
}

// Data-layer picks up on that action and requests a token using that payload.
// On success it dispatches:
{
  type: 'TOKEN_RECEIVE',
  token: '...tokendata...',
  signature: happychatTokenSignature
}

// Happychat middleware is listening for TOKEN_RECEIVE and when it sees
// one with a matching signature, it uses that token to connect chat. This also
// allows any other consumer to request tokens without impacting HC.
```

It's convenient in this case that the `signature` required to fetch a token could also be used as a unique identifier. In your case you may need to add some other uniquely-identifiable context that you can hook into from middleware (or even save into the store and let the component do the redirect on `didComponentUpdate()` checking the previous and current values from selectors).

These are just thoughts for if you decide to keep this in data-layer — @jordwest may be right that this stuff should be moved elsewhere (I'm honestly still a bit confused about what should and shouldn't go in here).